wire-ios-request-strategy/Sources/Helpers/FederationDeleteManager.swift

```diff
@@ -0,0 +1,139 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+
+import Foundation
+import WireDataModel
+
+final public class FederationDeleteManager {
+    private weak var syncContext: NSManagedObjectContext?
+
+    public init(syncContext: NSManagedObjectContext? = nil) {
+        self.syncContext = syncContext
+    }
+
+    public func backendStoppedFederatingWithDomain(domain: String) {
+        guard let moc = syncContext,
+              let selfDomain = ZMUser.selfUser(in: moc).domain else { return }
+
+        // search all conversations hosted on domain and remove participants from selfDomain
+        removeAllParticipantsFromDomain(selfDomain, inConversationsOnDomain: domain)
+
+        // search all conversation hosted on selfDomain that have participants from domain and remove them
+        removeAllParticipantsFromDomain(domain, inConversationsOnDomain: selfDomain)
+
+
+        // find all conversations where participants are from domain and selfDomain
+        let conversations = ZMConversation.existingConversationsHostedOnDomainDifferentThan(myDomain: selfDomain,
+                                                                                            otherDomain: domain,
+                                                                                            moc: moc)
+        for currentConversation in conversations {
+            removeParticipantsFromDomains(domains: [domain, selfDomain], inConversation: currentConversation)
+        }
+
+        markAllOneToOneConversationsAsReadOnly(forDomain: domain)
+
+        ignoreAllPendingConnectionRequests(fromDomain: domain)
+
+        do {
+            try moc.save()
+        } catch {}
+    }
+
+    public func domainsStoppedFederating(domains: [String]) {
+        guard let moc = syncContext else { return }
+        let conversations = ZMConversation.allGroupConversationWithSomeDomain(moc: moc)
+
+        for currentConversation in conversations {
+            if domains.contains(currentConversation.domain ?? "") {
+                // if conversation hosted on one of domains, then remove only users from the other domain
+                let domainsToRemove = domains.filter { $0 != currentConversation.domain ?? "" }
+                removeParticipantsFromDomains(domains: domainsToRemove, inConversation: currentConversation)
+            } else {
+                //r emove participants from both domains
+                guard currentConversation.containsParticipantsFromAllDomains(domains: domains) else { continue }
+                removeParticipantsFromDomains(domains: domains, inConversation: currentConversation)
+            }
```

Comment:
I think we can reuse the logic in the previous method. It looks like the general behavior is that when two domains stop federating, remove the users from both domains from all groups, but don't remove users hosted on the same domain as the conversation. Right?